### ISSUE TITLE
VobSub Interlacing Inconsistencies

### ISSUE THREAD

**Issue by timj** [ecdye/macSubtitleOCR#21]
Trying a different disk from #20 and whilst the image data looks close it seems there is a slight blurriness to it that is messing up the OCR.

eg

![subtitle_11](https://github.com/user-attachments/assets/65e8796c-09fd-46fe-bbfe-cdd053eeb337)

with subtools looking like:

<img width="352" alt="image" src="https://github.com/user-attachments/assets/3fabf551-34ff-4794-9730-1b0869dff78f">

and VLC:

<img width="730" alt="image" src="https://github.com/user-attachments/assets/31d46be6-67b0-4e89-a3b0-a10050862643">

and the macSubtitleOCR subtitle comes out as:

```
11
00:00:47,719 --> 00:00:54,884
and Im crewing Welverhe slashing
piles with his adamantine claws.
```

vs SubTools with Tesseract:

```
and I'm drlwlng Wolverlne slashlng
prices with hls adamantlne claws.
```

That's a "good" example. Some other ones are much less usable. SubTools is much better though.

Here are the VOBSUB files:

[fuzz.tgz](https://github.com/user-attachments/files/17320035/fuzz.tgz)

**Comment by timj** [comment]
Oh, and if I take that VLC image and ask Apple Preview to copy and paste the text I get:

> and I'm drawing Wolverine slashing prices with his adamantine claws.

which is perfect.

**Comment by timj** [comment]
I should have tried this before, but if I copy and paste the text of the slightly fuzzy image from macSubtitleOCR that also returns a perfect version of the text...

**Comment by ecdye** [comment]
> I should have tried this before, but if I copy and paste the text of the slightly fuzzy image from macSubtitleOCR that also returns a perfect version of the text...

Ok, that bit is very interesting. I wonder how they have their settings tuned, because presumably it uses the same api.

**Comment by ecdye** [comment]
I'll try and investigate this one tonight, hopefully it's a simple fix.

**Comment by timj** [comment]
Definitely interesting that you are getting a different answer. It's clear the image is fuzzier but there is hope!

**Comment by ecdye** [comment]
Hope indeed, I'm wondering if I should just give up on the idea behind implementing my own decoders and just use an external library like FFmpeg, but I just really don't want to have to have any external dependencies

**Comment by timj** [comment]
I agree that it's really nice that it's so easy to build with no library dependencies.

**Comment by ecdye** [comment]
Well, I added ffmpeg support for now, at least that should unlock the power of the OCR engine. I haven't given up on figuring out the decoding in swift, but it will take time. Let me know if you find any issues with the ffmpeg features.

**Comment by timj** [comment]
Ok. Thank you.  I've taken a quick look and:

* ffmpeg being at least an order of magnitude faster is impressive (and helps when experimenting) -- unsurprisingly many people working over many years results in a highly optimized decoder.
* Comparing the images shows me that your original decoder generates almost exactly the same images in most cases except for the fuzz one mentioned above where ffmpeg does generate a cleaner image.
* Looks like you fixed the end timestamps. Thank you. Maybe that's from ffmpeg?
* Somehow even though the images look almost identical for one of my earlier examples the OCR is different and not necessarily better with ffmpeg. Sometimes it's correct with the internal decoder and sometimes better with ffmpeg. I'm not sure how that's possible.

```
--- classic/track_0.srt 2024-10-11 07:41:49
+++ new/track_0.srt     2024-10-11 07:36:04
@@ -1,4548 +1,4538 @@
 1
-00:00:06,006 --> 00:00:07,708
+00:00:06,006 --> 00:00:07,735
 NARRATOR:
-In the criminal justice system
+In the criminall ustice system
 
 2
-00:00:07,807 --> 00:00:11,644
-the people are represented
-by two separate
-yet equally important groups,
+00:00:07,807 --> 00:00:11,676
+the people are representedi
+byntwo separate
+yet'equally important groups,
 
 3
-00:00:11,744 --> 00:00:13,547
-the pollee
+00:00:11,745 --> 00:00:13,576
+the police
 who investigate crime
 
 4
-00:00:13,647 --> 00:00:16,416
-and the distriet attorneys
-who prosecute the offenders.
+00:00:13,647 --> 00:00:16,445
+anditherd istrictlattorneys
+who prosecute the offenders
 
 5
-00:00:16,516 --> 00:00:18,853
-These are their stories.
+00:00:16,516 --> 00:00:18,006
+Theserarel theirs fories
 
 6
-00:00:20,152 --> 00:00:22,355
-MAN: Because I don't want
-her to hear us talking,
-that's why.
+00:00:20,152 --> 00:00:22,382
+MAN: Because lIdon t want
+her to hear us talking.,
+that siwhy..
```

**Comment by ecdye** [comment]
All very interesting. For the end timestamps, I actually just implemented a universal check to make sure they didn't go past the start of the next subtitle. Funnily enough in my testing it seems like the FFmpeg timestamps were actually less accurate in general than my decoder, I'm still not quite sure why. 

As for the OCR differences I am still not sure why those happen, I'm wondering the same thing. 

And yes it truly is an order of magnitude faster, it definitely helps that it is much more mature. I also have not taken much time to improve the efficiency of my decoder, which would be very helpful as my implementation was very cobbled together and not designed for efficiency. I've had to reverse engineer a lot because the documentation hasn't been very clear.
